pkg/services/oauthtoken/oauth_token.go

```diff
@@ -126,76 +135,90 @@ func (o *Service) HasOAuthEntry(ctx context.Context, usr identity.Requester) (*l
 }
 
 // TryTokenRefresh returns an error in case the OAuth token refresh was unsuccessful
-// It uses a singleflight.Group to prevent getting the Refresh Token multiple times for a given User
-func (o *Service) TryTokenRefresh(ctx context.Context, usr identity.Requester) error {
+// It uses a server lock to prevent getting the Refresh Token multiple times for a given User
+func (o *Service) TryTokenRefresh(ctx context.Context, usr identity.Requester) (*oauth2.Token, error) {
 	ctxLogger := logger.FromContext(ctx)
 
 	if usr == nil || usr.IsNil() {
 		ctxLogger.Warn("Can only refresh OAuth tokens for existing users", "user", "nil")
 		// Not user, no token.
-		return nil
+		return nil, nil
 	}
 
 	if !usr.IsIdentityType(claims.TypeUser) {
 		ctxLogger.Warn("Can only refresh OAuth tokens for users", "id", usr.GetID())
-		return nil
+		return nil, nil
 	}
 
 	userID, err := usr.GetInternalID()
 	if err != nil {
 		ctxLogger.Warn("Failed to convert user id to int", "id", usr.GetID(), "error", err)
-		return nil
+		return nil, nil
 	}
 
 	ctxLogger = ctxLogger.New("userID", userID)
 
 	lockKey := fmt.Sprintf("oauth-refresh-token-%d", userID)
-	if _, ok := o.cache.Get(lockKey); ok {
-		ctxLogger.Debug("Expiration check has been cached, no need to refresh")
-		return nil
+
+	lockTimeConfig := serverlock.LockTimeConfig{
+		MaxInterval: 30 * time.Second,
+		MinWait:     160 * time.Millisecond, // p95 for token refresh is ~800ms, so at least it should wait 800ms before it fails (5 retries with (at least) 160 ms)
+		MaxWait:     250 * time.Millisecond,
 	}
-	_, err, _ = o.singleFlightGroup.Do(lockKey, func() (any, error) {
-		ctxLogger.Debug("Singleflight request for getting a new access token", "key", lockKey)
+
+	retryOpt := func(attempts int) error {
+		if attempts < 5 {
+			return nil
+		}
+		return ErrCouldntRefreshToken
+	}
+
+	var newToken *oauth2.Token
+	var cmdErr error
+
+	lockErr := o.serverLock.LockExecuteAndReleaseWithRetries(ctx, lockKey, lockTimeConfig, func(ctx context.Context) {
+		ctxLogger.Debug("Serverlock request for getting a new access token", "key", lockKey)
 
 		authInfo, exists, err := o.HasOAuthEntry(ctx, usr)
 		if !exists {
 			if err != nil {
 				ctxLogger.Debug("Failed to fetch oauth entry", "error", err)
-			} else {
-				// User is not logged in via OAuth no need to check
-				o.cache.Set(lockKey, struct{}{}, maxOAuthTokenCacheTTL)
 			}
-			return nil, nil
+			return
 		}
 
-		_, needRefresh, ttl := needTokenRefresh(authInfo)
+		_, needRefresh := needTokenRefresh(authInfo)
```

Comment:
I think we need the current token here, and in case it doesn't need to be refreshed then we need to set `newToken` to the `currentToken`:
```
if !needRefresh {
    newToken = currentToken
    return
}
```
This is required for handling the scenario when there is an ongoing refresh (within the lock) and another node is waiting for the lock. If the newToken isn't set to the current (refreshed one) then it's gonna return with a nil `newToken`. Step-by-step: 
1. Node1 acquires the lock and starts the refresh
2. Node2 tries to acquires the lock
3. the token refresh is done 
4. the lock is released 
5. Node2 (which was waiting and retrying) acquires the lock
6. fetches the token from DB
7. `needTokenRefresh` will be false and it returns
8. `newToken` is nil

Wdyt? @DanCech @kalleep

Reply:
Yes, I think that makes sense